alf/algorithms/dsac_algorithm.py

```diff
@@ -0,0 +1,462 @@
+                    critic_quantiles * delta_tau.unsqueeze(1)).sum(-1)
+                idx = torch.min(
+                    critic_mean, dim=1)[1]  # [B] or [B, reward_dim]
+                critic_quantiles = critic_quantiles[torch.
+                                                    arange(len(idx)), idx]
+            else:
+                # Compute the min quantile distribution by taking a minimum value
+                # across all critic replicas for each quantile value
+                critic_std_quantiles = critic_quantiles.std(1, unbiased=True)
+                extra_info = critic_std_quantiles
+                if self._num_critic_replicas == 2:
+                    critic_quantiles = critic_quantiles.min(dim=1)[0]
+                else:
+                    critic_mean_quantiles = critic_quantiles.mean(1)
+                    critic_quantiles = critic_mean_quantiles - critic_std_quantiles
+
+        return critic_quantiles, critics_state, extra_info
+
+    def _critic_train_step(self, inputs: TimeStep, state: SacCriticState,
+                           rollout_info: SacInfo, action, tau_info: TauInfo):
+        bs = self._mini_batch_size
+        # Calculate the critics and value for both the current observation
+        # and next observation
+        # [B * (T-1), n_quantiles]
+        critics, critics_state, _ = self._compute_critics(
+            self._critic_networks,
+            inputs.observation[:-bs],
+            rollout_info.action[:-bs],
+            tau_info.tau_hat[:-bs],
+            state.critics,
+            replica_min=False)
+        # Extend with zeros so that the tensors can be correctly reshape to
+        # [T, B, ...]
+        zeros = torch.zeros_like(critics[:bs])
+        critics = torch.cat((critics, zeros), dim=0)
+
+        with torch.no_grad():
+            target_critics, target_critics_state, _ = self._compute_critics(
+                self._target_critic_networks, inputs.observation[bs:],
+                action[bs:], tau_info.next_tau_hat[bs:], state.target_critics,
+                tau_info.next_delta_tau[bs:])
+            # Prepend with zeros so that the tensors can be correctly reshape to
+            # [T, B, ...]
+            zeros = torch.zeros_like(target_critics[:bs])
+            target_critics = torch.cat((zeros, target_critics), dim=0)
+
+        target_critic = target_critics.detach()
+
+        state = SacCriticState(
+            critics=critics_state, target_critics=target_critics_state)
+        info = SacCriticInfo(critics=critics, target_critic=target_critic)
+
+        if self._debug_summaries and alf.summary.should_record_summaries():
+            with torch.no_grad():
+                critics_eval = self._compute_critics(
+                    self._critic_networks,
+                    inputs.observation[:-bs],
+                    rollout_info.action[:-bs],
+                    self._eval_tau,
+                    state.critics,
+                    replica_min=False)[0]
```

Comment:
The idea is to use a "fixed" delta_tau for evaluation rather than the randomly sampled one as used to compute ``critics`` for more debugging purpose. This does introduce extra computational cost, but only when ``debug_summaries=True``. I can summarize the computed critics instead if that makes better sense.